optuna/storages/_cached_storage.py

```diff
@@ -243,76 +200,25 @@ def get_best_trial(self, study_id: int) -> FrozenTrial:
     def set_trial_state_values(
         self, trial_id: int, state: TrialState, values: Optional[Sequence[float]] = None
     ) -> bool:
-        with self._lock:
-            cached_trial = self._get_cached_trial(trial_id)
-            if cached_trial is not None:
-                # When a waiting trial is updated to running, its `datetime_start` must be
-                # updated. However, a waiting trials is never cached so we do not have to account
-                # for this case.
-                assert cached_trial.state != TrialState.WAITING
-
-                self._check_trial_is_updatable(cached_trial)
-                ret = self._backend.set_trial_state_values(trial_id, state=state, values=values)
-
-                if values is not None:
-                    cached_trial.values = values
-                cached_trial.state = state
-                if cached_trial.state.is_finished():
-                    backend_trial = self._backend.get_trial(trial_id)
-                    cached_trial.datetime_complete = backend_trial.datetime_complete
-                return ret
-
-        ret = self._backend.set_trial_state_values(trial_id, state=state, values=values)
-        if (
-            ret
-            and state == TrialState.RUNNING
-            and trial_id in self._trial_id_to_study_id_and_number
-        ):
-            # Cache when the local thread pop WAITING trial and start evaluation.
-            with self._lock:
-                study_id, _ = self._trial_id_to_study_id_and_number[trial_id]
-                self._add_trials_to_cache(study_id, [self._backend.get_trial(trial_id)])
-                self._studies[study_id].owned_or_finished_trial_ids.add(trial_id)
-        return ret
+        return self._backend.set_trial_state_values(trial_id, state=state, values=values)
```

Comment:
Just for the record, I compared the speed when applying the following change, but found no performance improvement. So the current logic seems to be reasonable.

**Changes:**

```python
        ret = self._backend.set_trial_state_values(trial_id, state=state, values=values)
        if state.is_finished() and trial_id in self._trial_id_to_study_id_and_number:
            backend_trial = self._backend.get_trial(trial_id)
            study_id, trial_number = self._trial_id_to_study_id_and_number[trial_id]
            with self._lock:
                study = self._studies[study_id]
                study.trials[trial_number] = backend_trial
                study.finished_trial_ids.add(trial_id)
        return ret
```

**Benchmark script:**
https://gist.github.com/c-bata/fa8efc075e86a1f7e6dd6d83b85c1017

**Result:**

| Branch | `n_params=5` (mean ± std) (s) | `n_params=30` (mean ± std) (s) |
| --- | --- | --- |
| [PR 4631](https://github.com/optuna/optuna/pull/4631) | 10.1026 ± 0.1046 | 35.8279 ± 0.1306 |
| [cache-set-trial-state-values](https://github.com/c-bata/optuna/pull/7/commits/fb6391e5f46d7953959710fbb2d8d0fca27b7683) | 9.6567 ± 0.1974 | 35.8301 ± 0.1761 |
| main | 7.6167 ± 0.0959 | 32.8131 ± 0.1010 |
| v3.1.1 | 9.0216 ± 0.1590 | 42.7193 ± 0.1940 |

(Repeated 5 times per each)